documentation-support/src/main/java/tech/picnic/errorprone/documentation/BugPatternTestExtractor.java

```diff
@@ -0,0 +1,180 @@
+
+import static com.google.errorprone.matchers.Matchers.hasAnnotation;
+import static com.google.errorprone.matchers.Matchers.instanceMethod;
+import static com.google.errorprone.matchers.method.MethodMatchers.staticMethod;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.collect.ImmutableList;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.annotations.Immutable;
+import com.google.errorprone.annotations.Var;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ClassTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.util.TreeScanner;
+import com.sun.tools.javac.util.Context;
+import java.util.ArrayList;
+import java.util.List;
+import org.jspecify.annotations.Nullable;
+import tech.picnic.errorprone.documentation.BugPatternTestExtractor.BugPatternTestDocumentation;
+
+/**
+ * An {@link Extractor} that describes how to extract data from a test that tests a {@code
+ * BugChecker}.
+ */
+@Immutable
+final class BugPatternTestExtractor implements Extractor<BugPatternTestDocumentation> {
+  private static final Matcher<MethodTree> JUNIT_TEST_METHOD =
+      hasAnnotation("org.junit.jupiter.api.Test");
+
+  // XXX: Improve support for correctly extracting multiple sources from a single
+  // `{BugCheckerRefactoring,Compilation}TestHelper` test.
+  @Override
+  public BugPatternTestDocumentation extract(ClassTree tree, Context context) {
+    VisitorState state = VisitorState.createForUtilityPurposes(context);
+    CollectBugPatternTests scanner = new CollectBugPatternTests(state);
+
+    tree.getMembers().stream()
+        .filter(MethodTree.class::isInstance)
+        .map(MethodTree.class::cast)
+        .filter(m -> JUNIT_TEST_METHOD.matches(m, state))
+        .forEach(m -> scanner.scan(m, null));
+
+    String className = tree.getSimpleName().toString();
+    return new AutoValue_BugPatternTestExtractor_BugPatternTestDocumentation(
+        className.substring(0, className.lastIndexOf("Test")),
+        scanner.getIdentificationTests(),
+        scanner.getReplacementTests());
+  }
+
+  // XXX: Further improve the heuristics for determining what a BugPattern test is.
+  @Override
+  public boolean canExtract(ClassTree tree, VisitorState state) {
+    String className = tree.getSimpleName().toString();
+    if (!className.endsWith("Test")) {
+      return false;
+    }
+
+    ScanBugPatternTest scanBugPatternTest = new ScanBugPatternTest();
+    scanBugPatternTest.scan(tree, state);
+
+    String bugPatternName = className.substring(0, className.lastIndexOf("Test"));
+    return scanBugPatternTest.hasTestUsingClassInstance(bugPatternName);
+  }
+
+  private static final class ScanBugPatternTest extends TreeScanner<@Nullable Void, VisitorState> {
```

Comment:
The name of this `Scanner` can probably be improved.... The same holds for the `encounteredClasses` and `hasTestUsingClassInstance`.... Idk why, but really struggled with these names 🤔. Definitely open to suggestions 😄.